### PR TITLE
Adds misc cluster settings

### PR DESCRIPTION
**Pull request by peteralfonsi**
### Description
Adds cluster settings for tiered caching: 
- Feature flag enabling/disabling tiered caching, will eventually be removed
- Enable/disable for the disk tier (dynamic) 
- Max size of the disk tier
- Max size of the in-memory RBM keystore for the disk tier (dynamic)
- Min/max thread numbers for ehcache disk tier
- Ehcache disk tier write concurrency
- Ehcache disk tier disk segments
- Disk tier stale key cleanup threshold (dynamic, WIP, stitch together with Kiran's code)

For any of these settings to be accessible, the feature flag must be on. 

The disk tier is disabled by default. If dynamically enabled, the request cache will create a disk tier. If dynamically disabled, the request cache will stop using this disk tier, but not delete it unless the cache clear API is called. This behavior is consistent with enabling/disabling the heap tier. If the disk tier is then reenabled, it will use the disabled disk tier rather than creating a new one. 

Tested manually and with IT tests.

### Related Issues
Part of tiered caching

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [N/A] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: distribution/src/config/opensearch.yml [hunk 1]
```diff
@@ -127,3 +127,16 @@ ${path.logs}
...
+# opensearch.experimental.feature.tiered_caching.enabled: false
+#
+# Settings related to tiered caching:
+#index.requests.cache.tiered.disk.keystore.size: 0.05%
+#index.requests.cache.tiered.disk.stale_cleanup_threshold: 0.5
+#indices.requests.cache.tiered.disk.ehcache.min_threads: 2
+#indices.requests.cache.tiered.disk.ehcache.max_threads: 2
+#indices.requests.cache.tiered.disk.ehcache.write_concurrency: 2
+#indices.requests.cache.tiered.disk.ehcache.segments: 16
+#
```
Comment (kiranprakash154) [inline distribution/src/config/opensearch.yml]: should we remove these ?

File: server/src/main/java/org/opensearch/common/cache/tier/DiskTierTookTimePolicy.java [hunk 2]
```diff
@@ -29,8 +29,8 @@
  */
 public class DiskTierTookTimePolicy implements CacheTierPolicy<BytesReference> {
     public static final Setting<TimeValue> DISK_TOOKTIME_THRESHOLD_SETTING = Setting.positiveTimeSetting(
-        "index.requests.cache.disk.tooktime.threshold",
-        new TimeValue(10),
+        "indices.requests.cache.disk.tooktime.threshold",
+        TimeValue.ZERO,
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/DiskTierTookTimePolicy.java:33]: why the change of 10 to zero ?

File: server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java [hunk 3]
```diff
@@ -50,6 +52,52 @@
  * @param <V> The value type of cache entries
  */
 public class EhCacheDiskCachingTier<K, V> implements DiskCachingTier<K, V> {
+    public static final Setting<Double> REQUEST_CACHE_DISK_STALE_KEY_THRESHOLD = Setting.doubleSetting(
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java]: How is this different than the setting I have added ? - INDICES_REQUEST_CACHE_DISK_CLEAN_THRESHOLD in https://github.com/peteralfonsi/OpenSearch/pull/13/files#diff-a93623446c8ee94dffd709a37ec974ee8c27f0f9f31335f35e753ed802a8dc7bR244

File: server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java [hunk 4]
```diff
@@ -50,6 +52,52 @@
...
+    );
+
+    // Ehcache disk write minimum threads for its pool
+    public static final Setting<Integer> REQUEST_CACHE_DISK_MIN_THREADS = Setting.intSetting(
+        "indices.requests.cache.tiered.disk.ehcache.min_threads",
+        2,
+        1,
+        5,
+        Property.NodeScope
+    );
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java]: can u explain what is 2, 1, 5 ?

File: server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java [hunk 5]
```diff
@@ -50,6 +52,52 @@
...
+
+    // Ehcache disk write maximum threads for its pool
+    public static final Setting<Integer> REQUEST_CACHE_DISK_MAX_THREADS = Setting.intSetting(
+        "indices.requests.cache.tiered.disk.ehcache.max_threads",
+        2,
+        1,
+        20,
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java:71]: same

File: server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java [hunk 6]
```diff
@@ -50,6 +52,52 @@
...
+        Property.NodeScope
+    );
+
+    // Not be to confused with number of disk segments, this is different. Defines
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java:75]:
```suggestion
    // Not to be confused with number of disk segments, this is different. Defines
```

File: server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java [hunk 7]
```diff
@@ -50,6 +52,52 @@
...
+    // distinct write queues created for disk store where a group of segments share a write queue. This is
+    // implemented with ehcache using a partitioned thread pool exectutor By default all segments share a single write
+    // queue ie write concurrency is 1. Check OffHeapDiskStoreConfiguration and DiskWriteThreadPool.
+    public static final Setting<Integer> REQUEST_CACHE_DISK_WRITE_CONCURRENCY = Setting.intSetting(
+        "indices.requests.cache.tiered.disk.ehcache.write_concurrency",
+        2,
+        1,
+        3,
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java:83]: code comments on these

File: server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java [hunk 8]
```diff
@@ -50,6 +53,13 @@
  * The store estimates its memory footprint and will stop adding more values once it reaches its memory cap.
  */
 public class RBMIntKeyLookupStore implements KeyLookupStore<Integer> {
+    public static final Setting<ByteSizeValue> INDICES_CACHE_KEYSTORE_SIZE = Setting.memorySizeSetting(
+        "indices.requests.cache.tiered.disk.keystore.size",
+        "0.05%", // 5% of INDICES_CACHE_QUERY_SIZE
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java:58]: this should either be a 0.05 represent as 5% or 5%

File: server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java [hunk 9]
```diff
@@ -82,17 +92,23 @@ public int getValue() {
     protected final Lock readLock = lock.readLock();
     protected final Lock writeLock = lock.writeLock();
     private long mostRecentByteEstimate;
-    protected final int REFRESH_SIZE_EST_INTERVAL = 10000;
+    public static final int REFRESH_SIZE_EST_INTERVAL = 10000;
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java]:
```suggestion
    public static final int REFRESH_SIZE_EST_INTERVAL = 10_000;
```

Comment (kiranprakash154) [reply server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java]: is this made public static only for tests ? If yes, Can we avoid this ? i guess we can set a helper method with word `forTesting` in it, to set values to what we want accordingly.

File: server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java [hunk 10]
```diff
@@ -130,7 +147,7 @@ public boolean add(Integer value) {
         stats.numAddAttempts.inc();
 
         if (getSize() % REFRESH_SIZE_EST_INTERVAL == 0) {
-            mostRecentByteEstimate = getMemorySizeInBytes();
+            mostRecentByteEstimate = computeMemorySizeInBytes();
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java:150]: I like computeMemorySizeInBytes

File: server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java [hunk 11]
```diff
@@ -114,22 +148,17 @@ private EhCacheDiskCachingTier(Builder<K, V> builder) {
             this.threadPoolAlias = builder.threadPoolAlias;
         }
         this.settings = Objects.requireNonNull(builder.settings, "Settings objects shouldn't be null");
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java]: do we need settings anymore?

File: server/src/main/java/org/opensearch/common/cache/tier/TieredCacheSpilloverStrategyService.java [hunk 12]
```diff
@@ -193,6 +210,61 @@ private Function<K, CacheValue<V>> getValueFromTierCache(boolean trackStats) {
...
+        assert cachingTierList.size() == 1 && diskCachingTier.isPresent() && !diskTierInUse;
+        // list only contains heap tier, and there is an inactive disk tier we can reactivate
+        cachingTierList = List.of(onHeapCachingTier, diskCachingTier.get());
+        diskTierInUse = true;
+    }
+
+    /**
+     * Dynamically disable an existing disk tier.
+     */
+    private void disableDiskTier() {
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/common/cache/tier/TieredCacheSpilloverStrategyService.java:235]: i think we should return a boolean in these methods. Return false if something failed and true if succeeded. 
The caller can bubble it up as a failure or success. 
same for enableInactiveDiskTier, addNewDiskTier & toggleDiskTierEnabled
what do you think ?

File: server/src/main/java/org/opensearch/indices/IndicesRequestCache.java [hunk 13]
```diff
@@ -466,4 +496,31 @@ long count() {
...
         return registeredClosedListeners.size();
     }
+
+    /**
+     * Creates a new disk tier instance. Should only be run if the instance will actually be used!
+     * Otherwise, it may not be closed properly.
+     * @return A new disk tier instance
+     */
+    @Override
+    public EhCacheDiskCachingTier<Key, BytesReference> getDiskTier() {
```
Comment (kiranprakash154) [inline server/src/main/java/org/opensearch/indices/IndicesRequestCache.java]: getDiskTier method creating a new instance everytime could be misleading and I dont think we should rely on comments only to warn of its usage, its very easy to just call this without reading the code comments. 
How about we rename this to something that is self explanatory ? Like getNewDiskTier or something similar ?

File: server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java (hunk 3 above)
Comment (peteralfonsi) [reply server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java]: Yeah, we don't need this one I've added - didn't know you were implementing the cluster setting part too

File: server/src/main/java/org/opensearch/common/cache/tier/DiskTierTookTimePolicy.java (hunk 2 above)
Comment (peteralfonsi) [reply server/src/main/java/org/opensearch/common/cache/tier/DiskTierTookTimePolicy.java:33]: Oops - this was a change for the benchmark. This shouldn't be zero in the feature branch - we will see final value based on benchmark scenario 8

File: server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java (hunk 4 above)
Comment (peteralfonsi) [reply server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java]: Default/min/max - intellij shows it on my end, but i'll add a comment to explain too

File: server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java (hunk 8 above)
Comment (peteralfonsi) [reply server/src/main/java/org/opensearch/common/cache/tier/keystore/RBMIntKeyLookupStore.java:58]: The setting constructor takes in % of total heap size and we want 5% of 1% (IRC size). But i should probably just multiply 0.05 * the actual value of IRC max size from settings, to make this more flexible/clearer

File: server/src/main/java/org/opensearch/common/cache/tier/TieredCacheSpilloverStrategyService.java (hunk 12 above)
Comment (peteralfonsi) [reply server/src/main/java/org/opensearch/common/cache/tier/TieredCacheSpilloverStrategyService.java:235]: I like that idea

File: server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java (hunk 11 above)
Comment (peteralfonsi) [reply server/src/main/java/org/opensearch/common/cache/tier/EhCacheDiskCachingTier.java]: You're right i dont think we do

File: distribution/src/config/opensearch.yml (hunk 1 above)
Comment (peteralfonsi) [reply distribution/src/config/opensearch.yml]: Yeah, they might be too specific. I'll leave only the feature flag one in the comment
